discord/commands/mixins.py

```diff
@@ -0,0 +1,965 @@
+to deal in the Software without restriction, including without limitation
+the rights to use, copy, modify, merge, publish, distribute, sublicense,
+and/or sell copies of the Software, and to permit persons to whom the
+Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.
+"""
+
+from __future__ import annotations
+
+import asyncio
+import datetime
+import functools
+from typing import TYPE_CHECKING, Any, Callable, Coroutine, Generic, TypeVar, Union
+
+from .. import abc, utils
+from ..errors import (
+    ApplicationCommandError,
+    CheckFailure,
+    CommandError,
+    CommandInvokeError,
+    CommandOnCooldown,
+    DisabledCommand,
+)
+from .cooldowns import BucketType, CooldownMapping, MaxConcurrency, Cooldown
+
+if TYPE_CHECKING:
+    from typing_extensions import ParamSpec, Concatenate
+
+    from ..abc import MessageableChannel
+    from ..bot import AutoShardedBot, Bot
+    from ..cog import Cog
+    from ..guild import Guild
+    from ..interactions import Interaction
+    from ..member import Member
+    from ..message import Message
+    from ..state import ConnectionState
+    from ..user import ClientUser, User
+    from ..voice_client import VoiceProtocol
+
+    P = ParamSpec("P")
+else:
+    P = TypeVar("P")
+
+
+BotT = TypeVar("BotT", bound="Union[Bot, AutoShardedBot]")
+CogT = TypeVar("CogT", bound="Cog")
+
+T = TypeVar("T")
+Coro = Coroutine[Any, Any, T]
+Callback = Callable[Concatenate[CogT, "BaseContext", P], Coro[T]] | Callable[Concatenate["BaseContext", P], Coro[T]]
+MaybeCoro = Union[T, Coro[T]]
+
+Check = Union[
+    Callable[[CogT, "BaseContext"], MaybeCoro[bool]],
+    Callable[["BaseContext"], MaybeCoro[bool]],
+]
+
+Error = Union[
+    Callable[[CogT, "BaseContext[Any]", CommandError], Coro[Any]],
+    Callable[["BaseContext[Any]", CommandError], Coro[Any]],
+]
+ErrorT = TypeVar("ErrorT", bound="Error")
+
+Hook = Union[Callable[[CogT, "BaseContext"], Coro[Any]], Callable[["BaseContext"], Coro[Any]]]
+HookT = TypeVar("HookT", bound="Hook")
+
+
+__all__ = (
+    "Invokable",
+    "_BaseCommand",
+    "BaseContext",
+)
+
+
+def unwrap_function(function: functools.partial | Callable) -> Callback:
+    while True:
+        if hasattr(function, "__wrapped__"):
+            function = function.__wrapped__  # type: ignore # function may or may not have attribute
```

Comment:
Wdym "may or may not have attribute"? If it gets to this point, the function obviously has the attribute `__wrapped__`. You're probably getting a type checking error related to not knowing the type of `__wrapped__`, in that case, just cast it to a generic callable.

Reply:
As in like removing the functools.partial typehint?

Reply:
What? No.
Just use `typing.cast` on `function.__wrapped__`.

Reply:
Ooh wait, the type checking error is related to not knowing whether or not `__wrapped__` is an attribute to `functools.partial`:
```
Cannot access member "__wrapped__" for type "partial[Unknown]"
  Member "__wrapped__" is unknown
```

Reply:
For some reason, that's normal.
Use `getattr` instead.

Reply:
Even if you use `getattr`, you'd still need to go through the if statements

Reply:
Wdym?

Reply:
```py
while True:
    if isinstance(function, functools.partial):
        function = function.func
    elif wrapped := getattr(function, "__wrapped__", None):
        function = wrapped
    else:
        return function
```
for example, isn't that different

Reply:
And is that a bad thing?

Reply:
Nevermind, I was thinking of something else